DasharoPayloadPkg/Library/EfiVarsLib/EfiVarsLib.c

```diff
@@ -0,0 +1,251 @@
+  // Actual size of the storage is block size, the rest is
+  // Fault Tolerant Write (FTW) space and the FTW spare space.
+  VarStoreHdr->Size      = BlockSize - VolHdrLen;
+  VarStoreHdr->Format    = VARIABLE_STORE_FORMATTED;
+  VarStoreHdr->State     = VARIABLE_STORE_HEALTHY;
+
+  return TRUE;
+}
+
+BOOLEAN
+EFIAPI
+EfiVarsInit (
+  IN OUT MemRange  Fv,
+  OUT EfiVars      *Storage
+  )
+{
+  CONST EFI_FIRMWARE_VOLUME_HEADER  *VolHdr;
+  CONST VARIABLE_STORE_HEADER       *VarStoreHdr;
+  UINT8                             *VolData;
+  UINTN                             VolDataSize;
+
+  if (!InitFv (Fv)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a(): failed to create variable store",
+      __FUNCTION__
+      ));
+    return FALSE;
+  }
+
+  VolHdr      = (CONST EFI_FIRMWARE_VOLUME_HEADER *) Fv.Start;
+  VolData     = Fv.Start + VolHdr->HeaderLength;
+  VolDataSize = Fv.Length - VolHdr->HeaderLength;
+  VarStoreHdr = (CONST VARIABLE_STORE_HEADER *) VolData;
+
+  Storage->Fv           = Fv;
+  Storage->Store.Start  = VolData + sizeof (*VarStoreHdr);
+  Storage->Store.Length = VolDataSize - sizeof (*VarStoreHdr);
+  Storage->Vars         = NULL;
+
+  return TRUE;
+}
+
+STATIC
+VOID
+StoreVar (
+  IN  CONST EfiVar  *Var,
+  OUT UINT8         *Data
+  )
+{
+  AUTHENTICATED_VARIABLE_HEADER hdr;
+  SetMem (&hdr, sizeof (hdr), 0xff);
+
+  hdr.StartId    = VARIABLE_DATA;
+  hdr.State      = VAR_ADDED;
+  hdr.Reserved   = 0;
+  hdr.Attributes = Var->Attrs;
+  hdr.VendorGuid = Var->Guid;
+  hdr.NameSize   = Var->NameSize;
+  hdr.DataSize   = Var->DataSize;
+
+  CopyMem (Data, &hdr, sizeof (hdr));
+  Data += sizeof (hdr);
+
+  CopyMem (Data, Var->Name, Var->NameSize);
+  CopyMem (Data + Var->NameSize, Var->Data, Var->DataSize);
+}
+
+BOOLEAN
+EFIAPI
+EfiVarsWrite (
+  IN EfiVars  *Storage
+  )
+{
+  MemRange  VarStore;
+  UINT8     *OutData;
+  EfiVar    *Var;
+  UINTN     VarSize;
+
+  VarStore = Storage->Store;
+  OutData  = VarStore.Start;
+
+  for (Var = Storage->Vars; Var != NULL; Var = Var->Next) {
+    VarSize =
+      sizeof (AUTHENTICATED_VARIABLE_HEADER) + Var->NameSize + Var->DataSize;
+    if (OutData + VarSize > VarStore.Start + VarStore.Length) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "%a(): not enough space to serialize Variable Store (have 0x%x).\n",
+        __FUNCTION__,
+        VarStore.Length
+        ));
+      return FALSE;
+    }
+
+    StoreVar (Var, OutData);
+    OutData += HEADER_ALIGN (VarSize);
+  }
+
+  // The rest is "uninitialized".
+  SetMem (OutData, VarStore.Length - (OutData - VarStore.Start), 0xff);
+
+  return TRUE;
+}
+
+EfiVar *
+EFIAPI
+EfiVarsCreateVar (
+  IN EfiVars  *Storage
+  )
+{
+  EfiVar *new_var;
+  EfiVar *Var;
+
+  new_var = AllocatePool (sizeof (*new_var));
+  if (new_var == NULL)
+    return NULL;
+
+  SetMem (new_var, sizeof (*new_var), 0);
+  new_var->Attrs = EFI_VARIABLE_NON_VOLATILE
+                 | EFI_VARIABLE_BOOTSERVICE_ACCESS
+                 | EFI_VARIABLE_RUNTIME_ACCESS;
```

Comment:
Unless I'm missing something, this is unnecessary. The caller writes proper attributes, and they may be different than these.